rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,177 @@
+
+Registration flow:
+
+1. Client is configured by the user to use `oidc-jwt` joining with a specific provider. The client then uses the provider-specific logic to obtain a token.
+2. The client will call the `RegisterUsingToken` endpoint, specifying the name of the token, and providing the token that it has collected.
+3. The server will attempt to fetch the Token resource for the specified token.
+4. The server will check JWT header to ensure the `alg` is one we have allow-listed (RS[256, 384, 512])
+5. The server will check the `kid` of the JWT header, and obtain the relevant JWK from the cache or from the specified issuers well-known JWKS endpoint. It will then use the JWK to validate the token has been signed by the issuer.
+6. Other key claims of the JWT will be validated:
+  - Ensure the Issued At Time (iat) is in the past.
+  - Ensure the Expiry Time (exp) is in the future.
+7. The user's [configured Common Expression Language rule](#configuration) for the token will be evaluated against the claims, to ensure that the token is allowed to register with the Teleport cluster.
+8. Certificates will be generated for the client. The generated certificates will be non-renewable, as the client will proceed through the same steps to generate new certificates. This prevents exfiltratred credentials being used to repeatedly generate more credentials, maintaining access to the system.
+
+We will re-use `go-jose@v2` for validation of JWTs, since this library is already in use within Teleport.
+
+#### Caching JWKs
+
+Special attention should be given to the logic around caching the JWKs.
+
+We will cache these for two reasons:
+
+- Improve the performance of validating JWTs, as we will not need to make a HTTPS request to the issuer.
+- Improve the reliability, as we can validate JWTs even if the issuer is experiencing some downtime.
+- Reduce the impact of Teleport on an issuer. If onboarding a large number of nodes, we do not want to unduly place pressure on the issuer. 
+
+We should keep in mind the following considerations:
+
+- When we are presented with a JWT with a previously unseen `kid`, we should re-check the issuer's JWKs, as they may have begun issuing tokens with a new JWK.
+- We should ensure that the TTL of the cache entries is relatively short, as we want to allow an issuer to revoke a JWK that has been stolen.
+
+#### Configuration
+
+We will leverage the existing Token configuration kind as used by static tokens and IAM joining:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: my-gce-token
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  roles: [Node]
+  join_method: oidc-jwt
+  issuer_url: https://accounts.google.com
+  allow: claims.aud == "noah.teleport.sh" && claims.google.compute_engine.project_id == "my-project" && claims.google.compute_engine.instance_name == "an-instance"
```

Comment:
I'm not sure this would be appropriate, in my head there should be a one to one mapping of a token to an issuer_url, otherwise we will need to allow for connecting to potential an unlimited number of issuers to fetch all of their key sets in order to be able to validate the token. I can see a model where we have issuer_url in the top level configuration, and multiple `allow` entries though for the actual claim validation.

Reply:
I was assuming that you could get the issuer from the JWT, and only check that one, is that not the case?

Reply:
Hmm yeah, I suppose you could check the issuer and then filter down the allows to only the rules that list that issuer.

 I guess I just preferred the mental separation that a Token resource would only be for a particular issuer, but I suppose this way avoids adding more fields to the top level of the Token spec.

Reply:
The original case for having `allow:` be a list was allowing the token name to be baked into an EC2 AMI used across many accounts (listing the account IDs in the `allow` roles).

This seems less relevant for the OIDC usecase. I think it would be fine to add `issuer_url` (or whatever) at the top level if that would make more sense, my concern which prompted the comment is that `allow` is already a list, and in your current schema it's a single value.

Reply:
Sure, switching to a list for the allowed CEL expressions sounds good. I'll do that.

[eafab08](https://github.com/gravitational/teleport/pull/14414/commits/eafab08af066c66c8d8d2822f2acfe44e18b5893)